### ISSUE TITLE
Add AbortController.signal option to eventOptions Wrapper

### ISSUE THREAD

**Issue by tronicboy1** [lit/lit#3164]
### Which package is this a feature request for?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

Thank you everyone for your hard work and contributions to Lit, I LOVE this library.

Is it possible to add an option to pass an AbortController's signal to a Lit EventListener with the eventOptions wrapper?

I would like to do something like this:
```
@eventListener({ signal: this.#controller.signal })
#handleAbortClick: EventListener = () => { this.#abortDownload(); }
```

I looked into the source code and it looks like it could be possible, what do you think?

I have never contributed to an open source project before but if possible I would love to help.

Thank you,
Austin

### Alternatives and Workarounds

To remove an event listener I invoke a state change where a dummy event listener is added. This requires a lot of extra work.

```
render() {
  return html`<button @click=${this.addAbort ? this.#handleAbortClick : () => {}}>Stop</button>`
}
```

**Comment by sorvell** [comment]
Can you provide a little more context and description?

My understanding is that `[AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)` is useful since it can be used to cancel an in progress asynchronous action like a fetch. How would that make sense to be integrated with an event listener?

Thanks.

**Comment by thepassle** [comment]
Abort signals can be used on eventlisteners, too

**Comment by justinfagnani** [comment]
Yeah, there's a `signal` option here: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#parameters

I think it's reasonable to add this as we intend to support all options. The one hangup is how to get the right signal. Decorators are not in class body scope - they don't have access to the instance `this` - so this example won't work:

```ts
@eventListener({ signal: this.#controller.signal })
#handleAbortClick: EventListener = () => { this.#abortDownload(); }
```

because `this` will be undefined.

The inline options pattern would work:

```ts
render() {
  return html`<button @click=${{handleEvent: this.#handleAbortClick, signal: this.#controller.signal}}>Stop</button>`
}
```

For a decorator we might need a callback:

```ts
@eventListener({ signal: (el) => el.#controller.signal })
#handleAbortClick: EventListener = () => { this.#abortDownload(); }
```

But that callback will have issues with private fields...

**Comment by sorvell** [comment]
TIL... seems kind of odd but ok, probably useful.

Here's a [quick example](https://lit.dev/playground/#gist=8714dfd55281b3e5c972ebc65f0c5816) of the callback version.

We'd probably want to put this into the EventPart rather than purely in the decorator, but it's possible this way:

**Comment by kevinpschaaf** [comment]
Seems like at a minimum we should add `.signal` diffing to `shouldRemoveListener` here: https://github.com/lit/lit/blob/main/packages/lit-html/src/lit-html.ts#L2013-L2020

Beyond that, we could consider having `eventOptions` do the `addInitializer` approach Steve prototyped above, but as Justin points out that still that still won't be compatible with native private fields, so need a little more discussion on that.

**Comment by tronicboy1** [comment]
Any chance that using the new ECMAScript Decorators could address the binding issues of adding the signal option in a decorator?

If I am correct, you can now access the instance in method decorators with the stage 3 proposal. 

https://github.com/tc39/proposal-decorators#example-bound

However switching from the legacy decorators to the new decorators would probably be a huge refactoring if I am not mistaken.

**Comment by justinfagnani** [comment]
@tronicboy1 The problem isn't accessing the instance in the decorator implementation, the problem is that the `@eventOptions` decorator reuses all the options given on every host instance, but with a signal you probably need a different signal per host, so instead of providing a signal option, the user would have to provide a factory of some sort. That's just a slightly awkward API, and additional code-size cost in EventPart, and it would be specific to signal. Maybe we can generalize to make all options providable form a factory.

Re the private field issue, it's actually not one. Decorator expressions are in class body scope, so function expressions can access private fields. The only trick is that in the new decorators standard, initializers are run before field initializers, so you have to wait before accessing the private field. Here's an example:

https://www.typescriptlang.org/play?jsx=0#code/FAYw9gdgzgLgBABwE5gQUyTAnnAvHAHgGEAaOANQD4AKAMwC45qBLRogSj0rgFcIBrCGADuETrm7VgcGXBgBDJAHM0MNgBt5UKAEEQINNrBIAImnBJ5MYwBVFKmMTJUS02eAgw0ADzVwimtp6Bkam5sZWxkSQXr5OFJTi3ADebh6xMAB08gAmOQCSEMwwzPLqzABeGNRaWBAgcLR8ICWQTJypsnAA9N1wAOpocAC2PLBwwvLFvJ7M6nDytF5Ijcxo6jlwzEUlZZUYUHBIfG4y8pPTAAwA3KdwHlBg6miZ6mBKdNQwABbMUOzsW4yAC+gOAwNuoEChx0Mk6cAAAshUBhsNRqOsknB1pkAMQAI0U7Dc8n0hkeK1oYDAeDgAHJ5PiQHTITICYpaQAWABMt2BwFAkHG8lpEDQwjgOmoYKAA

**Comment by tronicboy1** [comment]
@justinfagnani 
Thank you for the excellent explanation! So basically if we go pass the `signal` instance in the `@eventOptions` decorator parameters it would have to be defined before the class, making it the same object for all instances... Very tricky! 

As you say a factory approach doesn't seem that bad from a user perspective, but wouldn't that involve breaking the current API? Or would it be optional to provide a function or object? 

`@eventOptions((el) => ({ signal: el.controller.signal }))` 

The issue with private fields in the new standard, thanks for pointing this out to me! Very unfortunate consequence.
